### PR TITLE
Add bucket notification support for NAS gateway

### LINKED ISSUE #6885: Minio events do not work when MINIO_ETCD_ENDPOINTS is set
When "MINIO_ETCD_ENDPOINTS" is set in a federation mode, Minio events cannot be added.

## Current Behavior
mc events add minio/bucket1 arn:minio:sqs::1:elasticsearch
mc: <ERROR> Cannot enable notification on the specified bucket. A header you provided implies functionality that is not implemented

## Your Environment
docker version
Client: Docker Engine - Community
Version: 18.09.0
API version: 1.39
Go version: go1.10.4
Git commit: 4d60db4
Built: Wed Nov 7 00:47:51 2018
OS/Arch: windows/amd64
Experimental: false

Server: Docker Engine - Community
Engine:
Version: 18.09.0
API version: 1.39 (minimum version 1.12)
Go version: go1.10.4
Git commit: 4d60db4
Built: Wed Nov 7 00:55:00 2018
OS/Arch: linux/amd64
Experimental: false

docker run minio/minio version
Version: 2018-11-22T02:51:56Z
Release-Tag: RELEASE.2018-11-22T02-51-56Z
Commit-ID: eddf468

mc.exe version
Version: 2018-11-06T01:12:20Z
Release-tag: RELEASE.2018-11-06T01-12-20Z
Commit-id: b827fac6356f80354a1cf61f43cf5e5b60bbb511

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Add bucket notification support for NAS gateway

## Motivation and Context
Fixes #6885

## Regression
No

## How Has This Been Tested?
Setup NAS gateway with [etcd](https://github.com/minio/minio/blob/master/docs/sts/etcd.md)
```
export MINIO_SECRET_KEY=minio123 
export MINIO_ACCESS_KEY=minio
export MINIO_ETCD_ENDPOINTS=http://localhost:2379 
minio gateway nas /tmp/test
```

In my example I am using NSQ notifications [here](https://docs.minio.io/docs/minio-bucket-notification-guide.html#NSQ), once configured restart server you can obtained ARN 
```
SQS ARNs:  arn:minio:sqs::1:nsq
```

```
~ mc mb myminio/certs
~ mc event add myminio/certs arn:minio:sqs::1:nsq
Successfully added arn:minio:sqs::1:nsq

~ mc event list myminio/certs/
arn:minio:sqs::1:nsq   s3:ObjectCreated:*,s3:ObjectRemoved:*,s3:ObjectAccessed:*   Filter:
```

You can setup another nas gateway at port `9001` to use the same configiration, and watch for events using `nsq_tail` command.
```
nsq_tail -nsqd-tcp-address 127.0.0.1:4150 -topic minio
2018/12/01 11:49:21 Adding consumer for topic: minio
2018/12/01 11:49:21 INF    1 [minio/tail577334#ephemeral] (127.0.0.1:4150) connecting to nsqd
{"EventName":"s3:ObjectCreated:Put","Key":"testbucet/hosts","Records":[{"eventVersion":"2.0","eventSource":"minio:s3","awsRegion":"","eventTime":"2018-12-01T19:49:32Z","eventName":"s3:ObjectCreated:Put","userIdentity":{"principalId":"minio"},"requestParameters":{"accessKey":"minio","region":"","sourceIPAddress":"127.0.0.1"},"responseElements":{"x-amz-request-id":"156C4D4EB6AD5629","x-minio-origin-endpoint":"http://127.0.0.1:9001"},"s3":{"s3SchemaVersion":"1.0","configurationId":"Config","bucket":{"name":"testbucet","ownerIdentity":{"principalId":"minio"},"arn":"arn:aws:s3:::testbucet"},"object":{"key":"hosts","size":268,"eTag":"6e5ad789aa1b14afcd57efc596408bb2","contentType":"application/octet-stream","userMetadata":{"content-type":"application/octet-stream"},"versionId":"1","sequencer":"156C4D4EB711036A"}},"source":{"host":"","port":"","userAgent":"Minio (linux; amd64) minio-go/v6.0.8 mc/2018-11-22T20:50:57Z"}}]}
{"EventName":"s3:ObjectCreated:Put","Key":"certs/hosts","Records":[{"eventVersion":"2.0","eventSource":"minio:s3","awsRegion":"","eventTime":"2018-12-01T19:49:40Z","eventName":"s3:ObjectCreated:Put","userIdentity":{"principalId":"minio"},"requestParameters":{"accessKey":"minio","region":"","sourceIPAddress":"127.0.0.1"},"responseElements":{"x-amz-request-id":"156C4D50996F6773","x-minio-origin-endpoint":"http://127.0.0.1:9000"},"s3":{"s3SchemaVersion":"1.0","configurationId":"Config","bucket":{"name":"certs","ownerIdentity":{"principalId":"minio"},"arn":"arn:aws:s3:::certs"},"object":{"key":"hosts","size":268,"eTag":"6e5ad789aa1b14afcd57efc596408bb2","contentType":"application/octet-stream","userMetadata":{"content-type":"application/octet-stream"},"versionId":"1","sequencer":"156C4D5099A0FFC2"}},"source":{"host":"","port":"","userAgent":"Minio (linux; amd64) minio-go/v6.0.8 mc/2018-11-22T20:50:57Z"}}]}
```

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :x: [more...](#6a018b02-1978-4098-b92d-3730ab654b02) |
<a name="6a018b02-1978-4098-b92d-3730ab654b02"></a>
### 6908-06e1d17/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30782
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 176cce77a958:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 42 seconds
(2/13) Running aws-sdk-go tests ... done in 0 seconds
(3/13) Running aws-sdk-java tests ... done in 2 seconds
(4/13) Running aws-sdk-php tests ... done in 42 seconds
(5/13) Running aws-sdk-ruby tests ... done in 6 seconds
(6/13) Running mc tests ... done in 31 seconds
(7/13) Running minio-dotnet tests ... done in 2 minutes and 44 seconds
(8/13) Running minio-go tests ... done in 25 seconds
(9/13) Running minio-java tests ... FAILED in 3 seconds
{
  "name": "minio-java",
  "function": "putObject(String bucketName, String objectName, String filename)",
  "args": "filename: 65MB",
  "duration": 1023,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code=XAmzContentSHA256Mismatch, message=The provided 'x-amz-content-sha256' header does not match what was computed., bucketName=null, objectName=null, resource=/minio-java-test-nlabgo/minio-java-test-3brtnv5, requestId=156C4DF76644E860, hostId=3L137)\nrequest={method=PUT, url=http://minikube:30782/minio-java-test-nlabgo/minio-java-test-3brtnv5?uploadId=6c7cc3ae-5571-41ec-ab23-8dba779f7482&partNumber=5, headers=Host: minikube:30782\nUser-Agent: Minio (amd64; amd64) minio-java/dev\nx-amz-content-sha256: f859b3b16a0b9f811b001c1aa5504d4212349a3d0c6537e0b21217f325bbb243\nx-amz-date: 20181201T200137Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20181201/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=400, headers=Accept-Ranges: bytes\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: Minio/DEVELOPMENT.2018-12-01T19-51-56Z (linux; amd64)\nVary: Origin\nX-Amz-Request-Id: 156C4DF76644E860\nX-Minio-Deployment-Id: \nX-Xss-Protection: 1; mode=block\nDate: Sat, 01 Dec 2018 20:01:37 GMT\nTransfer-Encoding: chunked\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:1134), io.minio.MinioClient.executePut(MinioClient.java:1316), io.minio.MinioClient.executePut(MinioClient.java:1338), io.minio.MinioClient.putObject(MinioClient.java:3394), io.minio.MinioClient.putObject(MinioClient.java:3516), io.minio.MinioClient.putObject(MinioClient.java:3046), io.minio.MinioClient.putObject(MinioClient.java:3083), FunctionalTest.putObject_test2(FunctionalTest.java:439), FunctionalTest.runTests(FunctionalTest.java:2788), FunctionalTest.main(FunctionalTest.java:2942)]"
}

Executed 8 out of 13 tests successfully.
```

**Review by vadmeste — APPROVED** [review]
LGTM & tested

File: cmd/gateway/nas/gateway-nas.go [hunk 1]
```diff
@@ -118,15 +118,5 @@ func (g *NAS) Production() bool {
...
-
-// IsNotificationSupported returns whether notifications are applicable for this layer.
-func (l *nasObjects) IsNotificationSupported() bool {
-	return false
-}
-
-// IsCompressionSupported returns whether compression is applicable for this layer.
-func (l *nasObjects) IsCompressionSupported() bool {
-	return false
+	minio.ObjectLayer
```
Comment (poornas) [inline cmd/gateway/nas/gateway-nas.go:121]: IsCompressionSupported() should still be false, if the gateways have different compression env settings the gateway would not behave consistently w.r.t compressed objects.

**Review by poornas — CHANGES_REQUESTED** [review]
Tested notification support - looks good to me.

File: cmd/gateway/nas/gateway-nas.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/gateway/nas/gateway-nas.go:121]: The deployment of Gateways is always consistent and it should be. This is similar to access and secrets if they don't use the same settings they will fail to work. Same as encryption vault settings etc.

I would say we should enable it, compression is implemented at handler layer and there is no reason to disable it.

Comment (poornas) [reply cmd/gateway/nas/gateway-nas.go:121]: Agree, perhaps this should be documented in NAS/ gateway docs. Approving this

**Review by poornas — APPROVED** [review]

**Pull request merged**